Libraries/LibWeb/HTML/OffscreenCanvas.cpp

```diff
@@ -0,0 +1,380 @@
+        // 1.Let element be the document element of global's associated Document.
+        auto* element = window.associated_document().document_element();
+        // 2. If element is not null :
+        if (element) {
+            // FIXME: 1. Set the inherited language of this to element's language.
+            // FIXME: 2. Set the inherited direction of this to element's directionality.
+        }
+    }
+
+    return realm.create<OffscreenCanvas>(realm, bitmap);
+}
+
+// https://html.spec.whatwg.org/multipage/canvas.html#dom-offscreencanvas
+OffscreenCanvas::OffscreenCanvas(JS::Realm& realm, NonnullRefPtr<Gfx::Bitmap> bitmap)
+    : EventTarget(realm)
+    , m_bitmap { move(bitmap) }
+{
+}
+
+OffscreenCanvas::~OffscreenCanvas() = default;
+
+WebIDL::ExceptionOr<void> OffscreenCanvas::transfer_steps(HTML::TransferDataHolder&)
+{
+    // FIXME: Implement this
+    dbgln("(STUBBED) OffscreenCanvas::transfer_steps(HTML::TransferDataHolder&)");
+    return {};
+}
+
+WebIDL::ExceptionOr<void> OffscreenCanvas::transfer_receiving_steps(HTML::TransferDataHolder&)
+{
+    // FIXME: Implement this
+    dbgln("(STUBBED) OffscreenCanvas::transfer_receiving_steps(HTML::TransferDataHolder&)");
+    return {};
+}
+
+HTML::TransferType OffscreenCanvas::primary_interface() const
+{
+    // FIXME: Implement this
+    dbgln("(STUBBED) OffscreenCanvas::primary_interface()");
+    return {};
+}
+
+WebIDL::UnsignedLong OffscreenCanvas::width() const
+{
+    return m_bitmap->size().width();
+}
+
+WebIDL::UnsignedLong OffscreenCanvas::height() const
+{
+    return m_bitmap->size().height();
+}
+
+void OffscreenCanvas::reset_context_to_default_state()
+{
+    m_context.visit(
+        [](GC::Ref<OffscreenCanvasRenderingContext2D>& context) {
+            context->reset_to_default_state();
+        },
+        [](GC::Ref<WebGL::WebGLRenderingContext>& context) {
+            context->reset_to_default_state();
+        },
+        [](GC::Ref<WebGL::WebGL2RenderingContext>& context) {
+            context->reset_to_default_state();
+        },
+        [](Empty) {
+            // Do nothing.
+        });
+}
+
+void OffscreenCanvas::set_new_bitmap_size(Gfx::IntSize new_size)
+{
+    m_bitmap = MUST(Gfx::Bitmap::create(Gfx::BitmapFormat::RGBA8888, Gfx::IntSize { new_size.width(), new_size.height() }));
+
+    m_context.visit(
+        [&](GC::Ref<OffscreenCanvasRenderingContext2D>& context) {
+            context->set_size(new_size);
+        },
+        [&](GC::Ref<WebGL::WebGLRenderingContext>& context) {
+            context->set_size(new_size);
+        },
+        [&](GC::Ref<WebGL::WebGL2RenderingContext>& context) {
+            context->set_size(new_size);
+        },
+        [](Empty) {
+            // Do nothing.
+        });
+}
+NonnullRefPtr<Gfx::Bitmap> OffscreenCanvas::bitmap() const
+{
+    return m_bitmap;
+}
+
+WebIDL::ExceptionOr<void> OffscreenCanvas::set_width(WebIDL::UnsignedLong value)
+{
+    if (value > 2147483647)
```

Comment:
where does this check come from? is there a spec step for setting the width?

Otherwise, this should be expressed in terms of AK::NumericLimits. or a call to AK::clamp.

Reply:
It was a copy paste from the same code in HTMLCanvasElement 😓  The spec doesnt mention it as far as I know